### ISSUE TITLE
Support `x.__class__ is T` type guards

### ISSUE THREAD

**Issue by shiftinv** [microsoft/pyright#3964]
**Is your feature request related to a problem? Please describe.**
As documented [here](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards), pyright supports lots of patterns for type narrowing using type guards (which is really cool!), including `type(x) is T`, but not `x.__class__ is T`.

**Describe the solution you'd like**
For new-style classes, `type(x)` and `x.__class__` should be identical, so it'd be nice if the latter was supported as well.

**Additional context**
```py
value: str | int = ...  # type: ignore

if value.__class__ is int:
    reveal_type(value)
if type(value) is int:
    reveal_type(value)
if isinstance(value, int):
    reveal_type(value)
```
```sh
$ pyright test.py
...
Assuming Python version 3.10
Assuming Python platform Linux
...
pyright 1.1.271
.../test.py
  .../test.py:4:17 - information: Type of "value" is "str | int"
  .../test.py:6:17 - information: Type of "value" is "int"
  .../test.py:8:17 - information: Type of "value" is "int"
0 errors, 0 warnings, 3 informations 
```

**Comment by erictraut** [comment]
Each type guard pattern requires very specific logic which adds to runtime analysis cost and support burden. We do occasionally add support for new type guard patterns but only if we have signal that it's a common pattern. The pattern you're suggesting is not something I have seen very often, if ever. Mypy doesn't support this type guard pattern either, which is further evidence that it's not commonly used. So I think it's unlikely that we could justify adding this.

As you point out, there is a more common equivalent that is supported. I recommend switching to the more common pattern if you want your code to work well with static type checking.

Another option is to use user-defined type guards as defined in [PEP 647](https://peps.python.org/pep-0647/).
